### ISSUE TITLE
Ent Sidekiq::Limiter.leaky: RuntimeError: can't add a new key into hash during iteration 

### ISSUE THREAD

**Issue by stirkac** [sidekiq/sidekiq#5939]
Ruby version: ruby 3.2.2 (2023-03-30 revision e51014f9c0)
Rails version: Rails 7.0.5
Sidekiq / Pro / Enterprise version(s): sidekiq-ent (~> 7.1.0)! / sidekiq-pro (~> 7.1.0)!

sidekiq.rb:

```
Sidekiq.configure_server do |config|
  config.redis = { ssl_params: { verify_mode: OpenSSL::SSL::VERIFY_NONE } }
  config.super_fetch!
  config.reliable_scheduler!
end

Sidekiq.configure_client do |config|
  config.redis = { ssl_params: { verify_mode: OpenSSL::SSL::VERIFY_NONE } }
end

Sidekiq::Client.reliable_push! unless Rails.env.test? || Rails.env.cypress?

Sidekiq::Limiter.configure do |config|
  config.errors << ActiveResource::TooManyRequests
end
```

Backtrace received:

```
RuntimeError: can't add a new key into hash during iteration
[GEM_ROOT]/gems/sidekiq-ent-7.1.0/lib/sidekiq-ent/limiter/leaky.rb:84 :in `<class:LeakyBucket>`
 [GEM_ROOT]/gems/sidekiq-ent-7.1.0/lib/sidekiq-ent/limiter/leaky.rb:12 :in `<module:Limiter>`
 [GEM_ROOT]/gems/sidekiq-ent-7.1.0/lib/sidekiq-ent/limiter/leaky.rb:4 :in `<module:Sidekiq>`
 [GEM_ROOT]/gems/sidekiq-ent-7.1.0/lib/sidekiq-ent/limiter/leaky.rb:3 :in `<main>`
unknown:? :in `unknown`
unknown:? :in `unknown`
  [GEM_ROOT]/gems/bootsnap-1.16.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:32 :in `require`
  [GEM_ROOT]/gems/zeitwerk-2.6.4/lib/zeitwerk/kernel.rb:38 :in `require`
 [GEM_ROOT]/gems/sidekiq-ent-7.1.0/lib/sidekiq-ent/limiter.rb:249 :in `leaky`
 [PROJECT_ROOT]/app/models/shop.rb:537 :in `shopify_rest_rate_limiter`
 [PROJECT_ROOT]/app/models/shop.rb:445 :in `within_rest_rate_limit`
 [PROJECT_ROOT]/app/services/fulfillment_creator_v2.rb:6 :in `call`
 [PROJECT_ROOT]/app/services/order_fulfiller.rb:18 :in `perform`
  [GEM_ROOT]/gems/sidekiq-7.1.1/lib/sidekiq/processor.rb:209 :in `execute_job`
  [GEM_ROOT]/gems/sidekiq-7.1.1/lib/sidekiq/processor.rb:177 :in `block (2 levels) in process`
  [GEM_ROOT]/gems/sidekiq-7.1.1/lib/sidekiq/middleware/chain.rb:180 :in `traverse`
  [GEM_ROOT]/gems/sidekiq-7.1.1/lib/sidekiq/middleware/chain.rb:183 :in `block in traverse`
  [GEM_ROOT]/gems/newrelic_rpm-9.2.2/lib/new_relic/agent/instrumentation/sidekiq/server.rb:25 :in `block in call`
  [GEM_ROOT]/gems/newrelic_rpm-9.2.2/lib/new_relic/agent/instrumentation/controller_instrumentation.rb:384 :in `perform_action_with_newrelic_trace`
  [GEM_ROOT]/gems/newrelic_rpm-9.2.2/lib/new_relic/agent/instrumentation/sidekiq/server.rb:20 :in `call`
  [GEM_ROOT]/gems/sidekiq-7.1.1/lib/sidekiq/middleware/chain.rb:182 :in `traverse`
  [GEM_ROOT]/gems/sidekiq-7.1.1/lib/sidekiq/middleware/chain.rb:183 :in `block in traverse`
  [GEM_ROOT]/gems/sentry-sidekiq-5.9.0/lib/sentry/sidekiq/sentry_context_middleware.rb:26 :in `call`
  [GEM_ROOT]/gems/sidekiq-7.1.1/lib/sidekiq/middleware/chain.rb:182 :in `traverse`
  [GEM_ROOT]/gems/sidekiq-7.1.1/lib/sidekiq/middleware/chain.rb:183 :in `block in traverse`
 [GEM_ROOT]/gems/sidekiq-pro-7.1.1/lib/sidekiq/batch/middleware.rb:58 :in `call`
  [GEM_ROOT]/gems/sidekiq-7.1.1/lib/sidekiq/middleware/chain.rb:182 :in `traverse`
  [GEM_ROOT]/gems/sidekiq-7.1.1/lib/sidekiq/middleware/chain.rb:183 :in `block in traverse`
 [GEM_ROOT]/gems/sidekiq-ent-7.1.0/lib/sidekiq-ent/limiter/middleware.rb:10 :in `call`
  [GEM_ROOT]/gems/sidekiq-7.1.1/lib/sidekiq/middleware/chain.rb:182 :in `traverse`
  [GEM_ROOT]/gems/sidekiq-7.1.1/lib/sidekiq/middleware/chain.rb:183 :in `block in traverse`
  [GEM_ROOT]/gems/sidekiq-7.1.1/lib/sidekiq/metrics/tracking.rb:26 :in `track`
  [GEM_ROOT]/gems/sidekiq-7.1.1/lib/sidekiq/metrics/tracking.rb:122 :in `call`
  [GEM_ROOT]/gems/sidekiq-7.1.1/lib/sidekiq/middleware/chain.rb:182 :in `traverse`
  [GEM_ROOT]/gems/sidekiq-7.1.1/lib/sidekiq/middleware/chain.rb:183 :in `block in traverse`
  [GEM_ROOT]/gems/sidekiq-7.1.1/lib/sidekiq/middleware/chain.rb:182 :in `traverse`
  [GEM_ROOT]/gems/sidekiq-7.1.1/lib/sidekiq/middleware/chain.rb:173 :in `invoke`
  [GEM_ROOT]/gems/sidekiq-7.1.1/lib/sidekiq/processor.rb:176 :in `block in process`
  [GEM_ROOT]/gems/sidekiq-7.1.1/lib/sidekiq/processor.rb:140 :in `block (6 levels) in dispatch`
  [GEM_ROOT]/gems/sidekiq-7.1.1/lib/sidekiq/job_retry.rb:114 :in `local`
  [GEM_ROOT]/gems/sidekiq-7.1.1/lib/sidekiq/processor.rb:139 :in `block (5 levels) in dispatch`
  [GEM_ROOT]/gems/sidekiq-7.1.1/lib/sidekiq/rails.rb:16 :in `block in call`
  [GEM_ROOT]/gems/activesupport-7.0.5/lib/active_support/execution_wrapper.rb:92 :in `wrap`
  [GEM_ROOT]/gems/activesupport-7.0.5/lib/active_support/reloader.rb:72 :in `block in wrap`
  [GEM_ROOT]/gems/activesupport-7.0.5/lib/active_support/execution_wrapper.rb:92 :in `wrap`
  [GEM_ROOT]/gems/activesupport-7.0.5/lib/active_support/reloader.rb:71 :in `wrap`
  [GEM_ROOT]/gems/sidekiq-7.1.1/lib/sidekiq/rails.rb:15 :in `call`
  [GEM_ROOT]/gems/sidekiq-7.1.1/lib/sidekiq/processor.rb:135 :in `block (4 levels) in dispatch`
  [GEM_ROOT]/gems/sidekiq-7.1.1/lib/sidekiq/processor.rb:270 :in `stats`
  [GEM_ROOT]/gems/sidekiq-7.1.1/lib/sidekiq/processor.rb:130 :in `block (3 levels) in dispatch`
  [GEM_ROOT]/gems/sidekiq-7.1.1/lib/sidekiq/job_logger.rb:13 :in `call`
  [GEM_ROOT]/gems/sidekiq-7.1.1/lib/sidekiq/processor.rb:129 :in `block (2 levels) in dispatch`
  [GEM_ROOT]/gems/sidekiq-7.1.1/lib/sidekiq/job_retry.rb:81 :in `global`
  [GEM_ROOT]/gems/sidekiq-7.1.1/lib/sidekiq/processor.rb:128 :in `block in dispatch`
  [GEM_ROOT]/gems/sidekiq-7.1.1/lib/sidekiq/job_logger.rb:39 :in `prepare`
  [GEM_ROOT]/gems/sidekiq-7.1.1/lib/sidekiq/processor.rb:127 :in `dispatch`
  [GEM_ROOT]/gems/sidekiq-7.1.1/lib/sidekiq/processor.rb:175 :in `process`
  [GEM_ROOT]/gems/sidekiq-7.1.1/lib/sidekiq/processor.rb:82 :in `process_one`
  [GEM_ROOT]/gems/sidekiq-7.1.1/lib/sidekiq/processor.rb:72 :in `run`
  [GEM_ROOT]/gems/sidekiq-7.1.1/lib/sidekiq/component.rb:10 :in `watchdog`
  [GEM_ROOT]/gems/sidekiq-7.1.1/lib/sidekiq/component.rb:19 :in `block in safe_thread`
  [GEM_ROOT]/gems/newrelic_rpm-9.2.2/lib/new_relic/agent/tracer.rb:434 :in `block (2 levels) in thread_block_with_current_transaction`
  [GEM_ROOT]/gems/newrelic_rpm-9.2.2/lib/new_relic/agent/tracer.rb:356 :in `capture_segment_error`
  [GEM_ROOT]/gems/newrelic_rpm-9.2.2/lib/new_relic/agent/tracer.rb:433 :in `block in thread_block_with_current_transaction`
```

Please let me know if there's any other useful information I can share, but it appears it's an issue that only started occurring on the latest version, we've never seen it before.

**Comment by mperham** [comment]
Fixed in 7.1.1, out now.

**Comment by stirkac** [comment]
that was faaaast 🚀
